### PR TITLE
SOCKS5 proxy for the Logstash connection

### PR DESCRIPTION
**Pull request by andrewkroh**
Adds the ability to use a SOCKS5 proxy on the Logstash output.

```
output:
  logstash:
    hosts: ["remote-host:5044"]
    proxy_url: socks5://user:password@socks5-proxy:2233
    #proxy_use_local_resolver: false
```

~~The documentation hasn't been updated yet. I'll do that in a separate PR.~~

### CODE REVIEW

File: libbeat/outputs/logstash/transport.go [hunk 1]
```diff
@@ -10,6 +10,7 @@ import (
 	"time"
 
 	"github.com/elastic/beats/libbeat/logp"
+	"golang.org/x/net/proxy"
```
Comment (ruflin) [inline libbeat/outputs/logstash/transport.go]: Small formatting issue

**Comment by ruflin** [comment]
Change LGTM but build is failing on travis (for libbeat). I bugs me that with 400 files the dependency is quite big, but I assume there is no way around it?

File: libbeat/outputs/logstash/transport.go (hunk 1 above)
Comment (andrewkroh) [inline libbeat/outputs/logstash/transport.go]: I separated the elastic packages from the third-party packages.

**Comment by andrewkroh** [comment]
I removed the unused packages from `golang.org/x/net`. 432 files to 40. Let's see if it still builds.

**Comment by ruflin** [comment]
@andrewkroh Looks good
@urso Could you also have a look at this one and merge it?

**Comment by monicasarbu** [comment]
Nice!

**Comment by urso** [comment]
we want to keep the test files?

e.g. vendor/golang.org/x/net/publicsuffix/table_test.go ~15k lines...

**Comment by andrewkroh** [comment]
@urso Good point, I removed the `_test.go` files.

**Comment by ruflin** [comment]
@andrewkroh Sounds like something we should do for all dependencies? Did you use a flag to do this?

**Comment by andrewkroh** [comment]
No glide flags for this. I just used `rm *_test.go`.

**Comment by ruflin** [comment]
Ok, that means we have to "manually" remove then again when we use glide update.

**Comment by urso** [comment]
I kinda started to use glide.yml mostly to mention the repo + pin the version, but copy required files myself.

File: libbeat/outputs/logstash/config.go [hunk 2]
```diff
@@ -11,6 +11,16 @@ type logstashConfig struct {
 	CompressionLevel int                `config:"compression_level"`
 	MaxRetries       int                `config:"max_retries"`
 	TLS              *outputs.TLSConfig `config:"tls"`
+	SOCKSProxy       *socksProxyConfig  `config:"socks_proxy"`
+}
+
+// socksProxyConfig holds the configuration information required to proxy
+// Logstash connections through a SOCKS5 server.
+type socksProxyConfig struct {
+	HostPort  string `config:"hostport"`   // host:port of the SOCKS5 server.
```
Comment (urso) [inline libbeat/outputs/logstash/config.go]: maybe change config to 'host' + add ability to set username/password right in address (like we do support for other  outputs)

Comment (andrewkroh) [inline libbeat/outputs/logstash/config.go]: How about I replace  `HostPort`, `Username`, and `Password` with `ProxyUrl` (`proxy_url` in config)? This is more consistent with how you set the proxy for ES (it's also `proxy_url`).

So you would write:

```
socks_proxy:
  proxy_url: socks5://username:password@mysocksproxy:1234
  remote_dns: true
```

Comment (andrewkroh) [inline libbeat/outputs/logstash/config.go]: I updated the PR. The config is now consistent with Elasticsearch (but requires a `socks5://` URL).

```
output:
  logstash:
    hosts: ["remote-host:5044"]
    proxy_url: socks5://user:password@socks5-proxy:2233
```

**Comment by andrewkroh** [comment]
This updated this PR with documentation and added the proxy options to the configuration files.

**Comment by tsg** [comment]
@elastic/beats (testing, testing, does this alias notifiy you?) It almost sounds like Godep was better for our needs? At least it did this cleanup automatically and we didn't have to manually delete the `.git` folders.

**Comment by ruflin** [comment]
@tsg For libbeat it is nice to have all the files inside, for the other dependencies agree. I like that we use now the vendor directory to follow the "golang" way, but lets see if something else then glide pops up. Its good that we know better and better what are requirements are.

**Comment by tsg** [comment]
@ruflin Godep now writes the dependencies in the `vendor/` folder by default. I kind of prefer the cleaner yaml file, though.

Good point about libbeat, we need it with full files for the community beats. Not sure if any of the tools can do that selectively.

**Pull request merged**